walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,231 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashMap;
 
 public final class FindMeetingQuery {
+
+  /**
+   * This method checks if an event contains no mandatory attendees using a MeetingRequest
+   * @param request MeetingRequest object which contains list of attendees
+   * @param event The Event object's attendees are compare with the request object
+   * @return boolean if the provide event object contains no mandatory attendees
+   */
+  private boolean hasNoMandatoryAttendees(MeetingRequest request, Event event){
+    return Collections.disjoint(event.getAttendees(), request.getAttendees());
+  }
+
+  /**
+   * This method checks if an event contains only optional Attendees using a MeetingRequest
+   * @param request MeetingRequest object which contains list of optional attendees
+   * @param event The Event object's attendees are compare with the request object
+   * @return boolean if the provide event object contains only optional attendees
+   */
+  private boolean hasOnlyOptionalAttendees(MeetingRequest request, Event event){
+
+    for (String attendee : event.getAttendees()){
+      if (request.getAttendees().contains(attendee)){
+        return false;
+      }
+      if (!(request.getOptionalAttendees().contains(attendee))){
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * This method returns the correct time ranges if optional attendees are taken into account
+   * @param optionalEvents List of events with only optional attendees
+   * @param timeRanges A list of available TimeRanges without taken into account optional events 
+   * @return A list of TimeRanges depending on the optionalEvents list and final size of timeRanges
+   */
+  private Collection<TimeRange> resolveOptionalEventConflicts(ArrayList<Event> optionalEvents, 
+        Collection<TimeRange> timeRanges){
+
+    Collection<TimeRange> backUpTimeRanges = new ArrayList<TimeRange>(timeRanges);
```

Comment:
I think you can probably eliminate `backUpTimeRanges` here. How about just checking if `removedTimeRanges` and `timeRanges` are the same size (which means you would have ended up removing all the time ranges)?